### DISCUSSION TITLE
FieldArray antipattern

### DISCUSSION THREAD

**Discussion by tadeas22** [jaredpalmer/formik#3006] [Q&A]
Hello

is not code like this `key={index}` in `map` function antipattern? I think it is https://reactjs.org/docs/lists-and-keys.html#keys and now i have issue [with this recommendation](https://formik.org/docs/api/fieldarray).

**Comment by johnrom** [comment]
Indexed keys is an antipattern when there is other information you can use to persist a unique key across renders regardless of the array's sort. If you have another identifier you can use, you should use it. Otherwise, the React document says directly below it:

> When you don’t have stable IDs for rendered items, you may use the item index as a key as a last resort:
> ```jsx
> const todoItems = todos.map((todo, index) =>
>   // Only do this if items have no stable IDs
>   <li key={index}>
>     {todo.text}
>   </li>
> );
> ```
> We don’t recommend using indexes for keys _if the order of items may change._ [emphasis mine]

If, in the Formik example, each friend had an ID, you could use `friend.friendId` as the key. However, when the docs insert a new key later, they are inserting a blank string. They'd have to generate a unique identifier and persist it instead of using a raw string for the value of `friend`, like:

```ts
type MyValues = { 
  friends: { key: string, name: string } 
};
// then:
values.friends.map((friend, index) => (
 <div key={friend.key}>
// and:
arrayHelpers.push({ key: generateDynamicFriendId(), name: '' });
```

I can keep going on new problems you will face, so you can see why the docs keep it simple. I'm open to adding a note to the docs with a link to the React lists and keys docs under the section explaining using objects as values: https://formik.org/docs/api/fieldarray#fieldarray-array-of-objects

In fact, we should change the object-based version to use a persistent key.

**Comment by johnrom** [comment]
Added an issue here: https://github.com/formium/formik/issues/3007
